src/OpenTelemetry/Trace/Configuration/OpenTelemetrySdk.cs

```diff
@@ -106,5 +100,35 @@ public static void EnableOpenTelemetry(Action<OpenTelemetryBuilder> configureOpe
 
             ActivitySource.AddActivityListener(listener);
         }
+
+        internal static void BuildSamplingParameters(
+            in ActivityCreationOptions<ActivityContext> options, out ActivitySamplingParameters samplingParameters)
+        {
+            ActivityContext parentContext = options.Parent;
+            if (options.Parent == default)
+            {
+                // Check if there is already a parent for the current span.
+                var parentSpan = Activity.Current;
+                if (parentSpan != null)
+                {
+                    parentContext = parentSpan.Context;
+                }
+            }
+
+            // This is not going to be the final traceId of the Activity (if one is created), however, it is
```

Comment:
I am not sure if this is correct (except when starting a brand new activity without parent)
If there is an active parent (explicit or Activity.Current), then the traceid for the parent will be used as traceid for the activity to be created.

If there was no parent, then the `ActivityTraceId.CreateRandom();` you pass to Sampling will be different from actual TraceId of the Activity created.

Reply:
The unit test you have validates this. Can we state that TraceId is different only for the Root one scenario, and for everything else, the traceid is the correct one of the activity to be created.

Reply:
Agreed: the behavior here is a bit tricky. For the `ProbabilityActivitySampler` itself, it works because as you said it respects the sampled flag. However, in principle, someone may want to write a sampler that does its selection based on the actual bits of the `traceId`, and the one received by the sampler would not be the actual `traceId` of the root span. In other words per OTel spec, it is possible to write a sampler that selects based on `traceId` bits, ignoring the sampled flag, and still sampling complete traces.

Reply:
Ideally, we should be able to pass the already generated `traceId` to the Activity to be created.

@tarekgh I remember that had already stated this issue in other comment thread. Seeing it concretely here makes me wonder if we could force the Activity to be created to use the same one passed to the sampler.

Reply:
Sorry I am not clear about the ask here. In the current APIs, you can create the parent context and pass it to ActivtySource.StartActivity, and we'll honor this context and the new Activity will be created with the trace Id passed in this parent context.

Reply:
@cijothomas @tarekgh Putting this together with https://github.com/open-telemetry/opentelemetry-dotnet/pull/702/files#r431881590 I think that we should consider if all work that is done here (`BuildSamplingParameters` method) should be done before calling the first `GetRequestedDataUsingContext` directly in `System.Diagnostics`.

The main downside that I see is that since the current `Activity` is only created if sampled we may need to add it to the `AsyncLocal` for the "current context" otherwise this will be creating a bunch of traceIds if there are nested activities not being sampled.

Reply:
Actually there is a bug with the code as it is: if the real "logical root" is not sampled it is possible that one of its "logical children" becomes sampled and it will show up as an incorrect "root". The proposal above should fix that.

Reply:
https://github.com/dotnet/runtime/pull/37185/files#diff-ccf594406736f7234e01bead593d4c8aR140 - @pjanotti Please check this PR. It should address our concern?

Reply:
Partially: it solves the handling of the parent in the case that the root span is sampled (see https://github.com/open-telemetry/opentelemetry-dotnet/pull/702/files#r432778065). However, it doesn't solve the case of not sampling the root/parent since `Activity.Current` is going to be null in such cases and the original traceId for the (not sampled) root span lost.

In order to make each step clear and easy to review I suggest the PR on dotnet/runtime (https://github.com/dotnet/runtime/pull/37185) to be merged as it is since it fixes a specific issue on its own and we address the other part separately.

/cc @cijothomas @tarekgh